File: samples/python-sample-project/src/services/repository.py
"""Generic repository interface and in-memory implementation."""

from abc import ABC, abstractmethod
from typing import Generic, TypeVar, Optional

T = TypeVar("T")
ID = TypeVar("ID")


class Repository(ABC, Generic[T, ID]):
    """Generic repository interface for data access."""

    @abstractmethod
    def find_by_id(self, entity_id: ID) -> Optional[T]:
        """Finds an entity by its identifier."""
        ...

    @abstractmethod
    def find_all(self) -> list[T]:
        """Returns all entities."""
        ...

    @abstractmethod
    def save(self, entity: T) -> T:
        """Saves an entity and returns the saved instance."""
        ...

    @abstractmethod
    def delete_by_id(self, entity_id: ID) -> None:
        """Deletes an entity by its identifier."""
        ...

    def count(self) -> int:
        """Returns the count of all entities."""
        return len(self.find_all())

File: samples/java-sample-project/src/main/java/com/example/services/Repository.java
package com.example.services;

import java.util.List;
import java.util.Optional;

/**
 * Generic repository interface for data access.
 *
 * @param <T> the entity type
 * @param <ID> the identifier type
 */
public interface Repository<T, ID> {

    /** Finds an entity by its identifier. */
    Optional<T> findById(ID id);

    /** Returns all entities. */
    List<T> findAll();

    /** Saves an entity and returns the saved instance. */
    T save(T entity);

    /** Deletes an entity by its identifier. */
    void deleteById(ID id);

    /** Returns the count of all entities. */
    default long count() {
        return findAll().size();
    }
}
